Documentation changes for #1940 (#1950)
metoppv/improver: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst

regimes and identify the associated distribution that describes the underlying sub-grid
variability.

Rainforests processes each ensemble member separately. The output produced from each ensemble member can be 
considered a pseudo-ensemble which represents a 
conditional probability distribution that describes the likelihood of observing a given outcome when the 

> realised atmospheric state is consistent with that represented in the input ensemble member forecast. 

Author: done

The predicted distributions for different ensemble members are then blended in probability space to produce 

> the final output.

Author: I've changed to "calibrated probability output", it's not really an ensemble.

identifying distinct weather types, the model bias and scale difference should be independent of any given location and time as 
the underlying physical process should be identical. Thus a grid-point can be calibrated using data from any location, provided the 
underlying weather type is the consistent. This enables effective calibration to be applied to areas that are typically lacking 
sufficient cases to calibrate against.

****************************
The RainForests method
****************************

RainForests is an adaptation of the ECPoint method that uses machine learning based
tree models to replace the manually constructed tree model of ECPoint. Here we use gradient-boosted
decision tree (GBDT) models.

Our aim is to produce a probability distribution of the expected rainfall, given the NWP 
forecasts of relevant variables. We define a set of rainfall thresholds, suitably spaced so as 
to accurately model the distribution. Then we train a separate GBDT model for each lead time and 
threshold (note that each GBDT model itself consists of several hundred trees).

> The input variables for the calibration are generally the NWP forecasts of the variable of interest, 
> and related variables (for example, when calibrating precipitation, a related variable would 
> be convective precipitation), along with ancillary variables such as solar time.

Reviewer: I think this is a good change, but I might suggest the following to be more focused:
```suggestion
The set of variables that feed into the GBDT models are what allows the calibration to distinguish between
different weather situations and build the appropriate probability distribution accordingly. These input
variables are typically diagnostic variables for the ensemble realization (including the variable of interest),
but can include static and dynamic ancillary variables, such as local solar time, and whole-of-ensemble
values for diagnostic variables, such as mean or standard-deviation.
```

Author: done

variability.

Rainforests (like ECPoint) processes each ensemble member separately to produce a per-realization output that can be
considered a pseudo-ensemble representing the
conditional probability distribution that describes the likelihood of observing a given outcome when the
realised atmospheric state is consistent with that represented in the input ensemble member forecast.
The predicted distributions for different ensemble members are then blended in probability space to produce 
the final calibrated probability output.

One advantage of the ECPoint approach is that the calibration is inherently non-local. As the calibration is done by
identifying distinct weather types, the model bias and scale difference should be independent of any given location and time as 
the underlying physical process should be identical. Thus a grid-point can be calibrated using data from any location, provided the 
underlying weather type is the consistent. This enables effective calibration to be applied to areas that are typically lacking 
sufficient cases to calibrate against.

****************************
The RainForests method
****************************

> RainForests is an adaptation of the ECPoint method that uses machine learning based
> tree models to replace the manually constructed tree model of ECPoint. Here we use gradient-boosted

Reviewer: ```suggestion
RainForests uses machine learning based tree models, namely gradient-boosted decision tree
(GBDT) models to replace the manually constructed decision tree model of ECPoint.
```

> decision tree (GBDT) models.

Reviewer: ```suggestion
```

Author: what did you mean here? it seems the suggested change just deleted the end of the sentence...

Reviewer: Sorry, I had another suggestion above this that hasn't seemed to have made it in. I'll add that comment in too... given this straddles both additions and removals, I was trying to frame the suggestions so that they could easily slot in (but I obviously failed).

Our aim is to produce a probability distribution of the expected rainfall, given the NWP 
forecasts of relevant variables. We define a set of rainfall thresholds, suitably spaced so as 
to accurately model the distribution. Then we train a separate GBDT model for each lead time and 
threshold (note that each GBDT model itself consists of several hundred trees).

The set of variables that feed into the GBDT models are what allows the calibration to distinguish between
different weather situations and build the appropriate probability distribution accordingly. These input
variables are typically diagnostic variables for the ensemble realization (including the variable of interest),
but can include static and dynamic ancillary variables, such as local solar time, and whole-of-ensemble
values for diagnostic variables, such as mean or standard deviation.

> We use LightGBM for training the models, and compile the models with Treelite for efficient prediction.

Reviewer: ```suggestion
Here we use LightGBM for training the models, and compile the models with Treelite for efficient prediction.
```

Author: done

The model training process is relatively simple and involves collating a series of
forecast-observation pairs with the associated feature variables into a single pandas

> dataframe. In general, each ensemble member yields a separate row of the dataframe (although for 
> reasons of computational efficiency it may be desirable to only use a subset of members, for example 
> by using only the control realization in each ensemble). 

Reviewer: I thin k this is much better. I'd recommend dropping the brackets:
```suggestion
dataframe. In general, each ensemble member yields a separate row of the dataframe, although for 
reasons of computational efficiency it may be desirable to only use a subset of members, such as the
control realization only.
```

Author: done

> Each model predicts the probability that rainfall will exceed a particular threshold, 
> so the output is expected to be a number between 0 and 1. The Improver code will cap model predictions outside 
> this range so they lie between 0 and 1. (This is necessary because we are currently training using the mean 
> squared error loss, there are occasional cases where the LightGBM model predicts vlaues that are slightly more than 1 
> or less than 0.)

Reviewer: I think this is also a good update, but I might suggest rewording slightly:
```suggestion
As each model predicts the probability that rainfall will exceed a particular threshold, 
the output is expected to be a number between 0 and 1. However, on occasion the GBDT models may
predict values slightly outside of this range and so the model predictions are capped to back onto this
interval; such values are possible due to the current choice of loss function used in training, namely the
mean squared loss function which is akin to Brier score when working in probability space.
```

Author: done

> Currently model training is done offline.

Reviewer: Is there are reason to drop the comment about the length of the training period?

Author: not really, restored it now

feature cubes. The tree-models are passed in via a model-config json which identifies
the appropriate tree-model file for each error-threshold.

The decision-tree models are used to construct representative probability distributions for
each input ensemble member which are then blended to give the calibrated
distribution for the full ensemble.

The distributions for the individual ensemble members are formed in a two-step process:

> 1. Evaluate CDF defined over the specified model thresholds for each ensemble member.

Reviewer: 1. Evaluate the CDF defined over the specified model thresholds for each ensemble member.

Author: done

feature cubes. The tree-models are passed in via a model-config json which identifies

> the appropriate tree-model file for each error-threshold.

Reviewer: ```suggestion
the appropriate tree-model file for each threshold.
```

Author: done

The decision-tree models are used to construct representative probability distributions for
each input ensemble member which are then blended to give the calibrated
distribution for the full ensemble.

The distributions for the individual ensemble members are formed in a two-step process:

1. Evaluate the CDF defined over the specified model thresholds for each ensemble member.
Each threshold exceedance probability is evaluated using the corresponding
decision-tree model.

> 2. Interpolate each ensemble member distribution to the output thresholds.

Reviewer: The formatting in the markdown seems a bit off. Might be worth double checking what's causing the misalignment (maybe the extra line)?

Author: fixed, it was the extra line
